// @bryntum/calendar/source/lib/Calendar/localization/Ar.js
import LocaleHelper from '../../Core/localization/LocaleHelper.js';
import '../../Scheduler/localization/Ar.js';

const locale = {

    localeName : 'Ar',
    localeDesc : 'اللغة العربية',
    localeCode : 'ar',

    EventEdit : {
        Calendar  : 'التقويم',
        'All day' : 'طوال اليوم',
        day       : 'يوم',
        week      : 'أسبوع',
        month     : 'شهر',
        year      : 'سنة',
        decade    : 'عقد'
    },

    EventMenu : {
        duplicateEvent : 'حدث مكرر',
        copy           : 'نسخ'
    },

    Calendar : {
        Today      : 'اليوم',
        next       : range => `${range} التالي`,
        previous   : range => `${range} السابق`,
        plusMore   : value => `+${value} أكثر`,
        allDay     : 'طوال اليوم',
        endsOn     : d => `${d} النهايات`,
        weekOfYear : ([y, w]) => `${y} ,${w} أسبوع`,
        loadFail   : 'فشل تحميل بيانات التقويم. يُرجى الاتصال بمسؤول النظام الخاص بك'
    },

    CalendarDrag : {
        holdCtrlForRecurrence : 'اضغط باستمرار على CTRL من أجل حدث متكرر'
    },

    CalendarMixin : {
        eventCount : count => `حدث${(count && count > 1) ? 'الأ' : ''} ${count || 'لا'}`
    },

    EventTip : {
        'Edit event' : 'تعديل حدث',
        timeFormat   : 'LST'
    },

    ModeSelector : {
        includeWeekends : 'تضمين عطلات نهاية الأسبوع',
        weekends        : 'عطلات نهاية الأسبوع'
    },

    AgendaView : {
        Agenda : 'جدول أعمال'
    },

    MonthView : {
        Month     : 'شهر',
        monthUnit : 'شهر'
    },

    WeekView : {
        weekUnit : 'أسبوع'
    },

    YearView : {
        Year     : 'سنة',
        yearUnit : 'سنة'
    },

    EventList : {
        List   : 'قائمة',
        Start  : 'بدء',
        Finish : 'إنهاء'
    },

    DayView : {
        Day               : 'يوم',
        dayUnit           : 'يوم',
        daysUnit          : 'أيام',
        expandAllDayRow   : 'توسيع قسم طوال اليوم',
        collapseAllDayRow : 'تصغير قسم طوال اليوم',
        timeFormat        : 'LST'
    },

    Sidebar : {
        'Filter events' : 'تصفية الأحداث'
    },

    WeekExpander : {
        expandTip   : 'انقر لتوسيع صف',
        collapseTip : 'انقر لتصغير صف'
    }
};

export default LocaleHelper.publishLocale(locale);
