// @bryntum/calendar/source/lib/Calendar/localization/Cs.js
import LocaleHelper from '../../Core/localization/LocaleHelper.js';
import '../../Scheduler/localization/Cs.js';

const locale = {

    localeName : 'Cs',
    localeDesc : 'Česky',
    localeCode : 'cs',

    EventEdit : {
        Calendar  : 'Kalendář',
        'All day' : 'Celý den',
        day       : 'Den',
        week      : 'Týden',
        month     : 'Měsíc',
        year      : 'Rok',
        decade    : 'Desetiletí'
    },

    EventMenu : {
        duplicateEvent : 'Duplikovat událost',
        copy           : 'kopírovat'
    },

    Calendar : {
        Today      : 'Dnes',
        next       : range => `Další ${range}`,
        previous   : range => `Předchozí ${range}`,
        plusMore   : value => `+${value} další`,
        allDay     : 'Celý den',
        endsOn     : d => `Končí ${d}`,
        weekOfYear : ([y, w]) => `Týden ${w}, ${y}`,
        loadFail   : 'Načtení dat kalendáře se nezdařilo. Obraťte se na správce systému'
    },

    CalendarDrag : {
        holdCtrlForRecurrence : 'Při opakující se události podržte CTRL'
    },

    CalendarMixin : {
        eventCount : count => {
            if (!count) {
                return 'Žádná událost';
            }
            else {
                const lastDigit = count[count.length - 1];
                const suffix = { 1 : '', 2 : 'i', 3 : 'i', 4 : 'i' }[lastDigit] || 'í';

                return count + ' událost' + suffix;
            }
        }
    },

    EventTip : {
        'Edit event' : 'Upravit událost',
        timeFormat   : 'LST'
    },

    ModeSelector : {
        includeWeekends : 'Zahrnout víkendy',
        weekends        : 'Víkendy'
    },

    AgendaView : {
        Agenda : 'Program'
    },

    MonthView : {
        Month     : 'Měsíc',
        monthUnit : 'měsíc'
    },

    WeekView : {
        weekUnit : 'týden'
    },

    YearView : {
        Year     : 'Rok',
        yearUnit : 'rok'
    },

    EventList : {
        List   : 'Seznam',
        Start  : 'Začátek',
        Finish : 'Konec'
    },

    DayView : {
        Day               : 'Den',
        dayUnit           : 'den',
        daysUnit          : 'dny',
        expandAllDayRow   : 'Rozšířit výběr na všechny dny',
        collapseAllDayRow : 'Sloučit výběr na všechny dny',
        timeFormat        : 'LST'
    },

    Sidebar : {
        'Filter events' : 'Filtrovat události'
    },

    WeekExpander : {
        expandTip   : 'Kliknutím rozšíříte řádek',
        collapseTip : 'Kliknutím sloučíte řádek'
    }
};

export default LocaleHelper.publishLocale(locale);
