// @bryntum/calendar/source/lib/Calendar/localization/Da.js
import LocaleHelper from '../../Core/localization/LocaleHelper.js';
import '../../Scheduler/localization/Da.js';

const locale = {

    localeName : 'Da',
    localeDesc : 'Dansk',
    localeCode : 'da',

    EventEdit : {
        Calendar  : 'Kalender',
        'All day' : 'Hele dag',
        day       : 'dagen',
        week      : 'Uge',
        month     : 'Måned',
        year      : 'År',
        decade    : 'Årti'
    },

    EventMenu : {
        duplicateEvent : 'Dubleret begivenhed',
        copy           : 'kopi'
    },

    Calendar : {
        Today      : 'I dag',
        next       : range => `Næste ${range}`,
        previous   : range => `Tidligere ${range}`,
        plusMore   : value => `+${value} mere`,
        allDay     : 'Hele dagen',
        endsOn     : d => `Slutter ${d}`,
        weekOfYear : ([y, w]) => `Uge ${w}, ${y}`,
        loadFail   : 'Indlæsning af kalenderdata mislykkedes. Kontakt venligst din systemadministrator'
    },

    CalendarDrag : {
        holdCtrlForRecurrence : 'Hold CTRL nede for en tilbagevendende begivenhed'
    },

    CalendarMixin : {
        eventCount : count => `${count || 'Ingen'} begivenhed${(count && count > 1) ? 'er' : ''}`
    },

    EventTip : {
        'Edit event' : 'Rediger begivenhed',
        timeFormat   : 'LST'
    },

    ModeSelector : {
        includeWeekends : 'Include weekends',
        weekends        : 'Weekender'
    },

    AgendaView : {
        Agenda : 'Dagsorden'
    },

    MonthView : {
        Month     : 'måned',
        monthUnit : 'måned'
    },

    WeekView : {
        weekUnit : 'uge'
    },

    YearView : {
        Year     : 'år',
        yearUnit : 'år'
    },

    EventList : {
        List   : 'Liste',
        Start  : 'Start',
        Finish : 'Afslut'
    },

    DayView : {
        Day               : 'dag',
        dayUnit           : 'dag',
        daysUnit          : 'dage',
        expandAllDayRow   : 'Udvid heldagsafsnittet',
        collapseAllDayRow : 'Skjul heldagsafsnittet',
        timeFormat        : 'LST'
    },

    Sidebar : {
        'Filter events' : 'Filtrer begivenheder'
    },

    WeekExpander : {
        expandTip   : 'Klik for at udvide rækken',
        collapseTip : 'Klik for at skjule rækken'
    }
};

export default LocaleHelper.publishLocale(locale);
