// @bryntum/calendar/source/lib/Calendar/localization/De.js
import LocaleHelper from '../../Core/localization/LocaleHelper.js';
import '../../Scheduler/localization/De.js';

const locale = {

    localeName : 'De',
    localeDesc : 'Deutsch',
    localeCode : 'de-DE',

    EventEdit : {
        Calendar  : 'Kalendar',
        'All day' : 'Ganzt채gig',
        day       : 'Tag',
        week      : 'Woche',
        month     : 'Monat',
        year      : 'Jahr',
        decade    : 'Jahrzehnt'
    },

    EventMenu : {
        duplicateEvent : 'Ereignis duplizieren',
        copy           : 'kopieren'
    },

    Calendar : {
        Today      : 'Heute',
        next       : range => `N채chste ${range}`,
        previous   : range => `Vorherige ${range}`,
        plusMore   : value => `+${value} mehr`,
        allDay     : 'Ganzt채gig',
        endsOn     : d => `Endet ${d}`,
        weekOfYear : ([y, w]) => `Woche ${w}, ${y}`,
        loadFail   : 'Kalenderdaten konnten nicht geladen werden. Bitte wenden Sie sich an Ihren Systemadministrator'
    },

    CalendarDrag : {
        holdCtrlForRecurrence : 'CTRL f체r ein wiederkehrendes Ereignis halten'
    },

    CalendarMixin : {
        eventCount : count => `${count || 'Kein'} Ereignis${(count && count > 1) ? 'sen' : ''}`
    },

    EventTip : {
        'Edit event' : 'Ereignis bearbeiten',
        timeFormat   : 'LST'
    },

    ModeSelector : {
        includeWeekends : 'Wochenenden einbeziehen',
        weekends        : 'Wochenenden'
    },

    AgendaView : {
        Agenda : 'Agenda'
    },

    MonthView : {
        Month     : 'Monat',
        monthUnit : 'monat'
    },

    WeekView : {
        weekUnit : 'woche'
    },

    YearView : {
        Year     : 'Jahr',
        yearUnit : 'jahr'
    },

    EventList : {
        List   : 'Liste',
        Start  : 'Start',
        Finish : 'Ende'
    },

    DayView : {
        Day               : 'Tag',
        dayUnit           : 'tag',
        daysUnit          : 'tage',
        expandAllDayRow   : 'Ganztagsbereich erweitern',
        collapseAllDayRow : 'Ganztagsbereich einklappen',
        timeFormat        : 'LST'
    },

    Sidebar : {
        'Filter events' : 'Ereignisse filtern'
    },

    WeekExpander : {
        expandTip   : 'Zum Erweitern der Zeile klicken',
        collapseTip : 'Zum Einklappen der Zeile klicken'
    }
};

export default LocaleHelper.publishLocale(locale);
