// @bryntum/calendar/source/lib/Calendar/localization/El.js
import LocaleHelper from '../../Core/localization/LocaleHelper.js';
import '../../Scheduler/localization/El.js';

const locale = {

    localeName : 'El',
    localeDesc : 'Ελληνικά',
    localeCode : 'el',

    EventEdit : {
        Calendar  : 'Ημερολόγιο',
        'All day' : 'Ολοήμερο',
        day       : 'Ημέρα',
        week      : 'Εβδομάδα',
        month     : 'Μήνας',
        year      : 'Έτος',
        decade    : 'Δεκαετία'
    },

    EventMenu : {
        duplicateEvent : 'Αντιγραφή συμβάντος',
        copy           : 'αντιγραφή'
    },

    Calendar : {
        Today      : 'Σήμερα',
        next       : range => `Επόμενο/η ${range}`,
        previous   : range => `Προηγούμενο/η ${range}`,
        plusMore   : value => `+${value} παραπάνω`,
        allDay     : 'Ολοήμερο',
        endsOn     : d => `Τελειώνει στις ${d}`,
        weekOfYear : ([y, w]) => `Εβδομάδα ${w}, ${y}`,
        loadFail   : 'Η φόρτωση δεδομένων ημερολογίου απέτυχε. Επικοινωνήστε με το διαχειριστή του συστήματός σας.'
    },

    CalendarDrag : {
        holdCtrlForRecurrence : 'Για επαναλαμβανόμενο συμβάν, κρατήστε πατημένο το πλήκτρο CTRL'
    },

    CalendarMixin : {
        eventCount : count => `${count || 'Κανένα'} συμβάν${(count && count > 1) ? 'εις' : ''}`
    },

    EventTip : {
        'Edit event' : 'Επεξεργασία συμβάντος',
        timeFormat   : 'LST'
    },

    ModeSelector : {
        includeWeekends : 'Συμπερίληψη Σαββατοκύριακων',
        weekends        : 'Σαββατοκύριακα'
    },

    AgendaView : {
        Agenda : 'Πρόγραμμα'
    },

    MonthView : {
        Month     : 'Μήνας',
        monthUnit : 'μήνας'
    },

    WeekView : {
        weekUnit : 'εβδομάδα'
    },

    YearView : {
        Year     : 'Έτος',
        yearUnit : 'έτος'
    },

    EventList : {
        List   : 'Λίστα',
        Start  : 'Αρχή',
        Finish : 'Τέλος'
    },

    DayView : {
        Day               : 'Ημέρα',
        dayUnit           : 'ημέρα',
        daysUnit          : 'ημέρες',
        expandAllDayRow   : 'Επεκτείνετε την ενότητα του ολοήμερου',
        collapseAllDayRow : 'Καταργήστε την ενότητα του ολοήμερου',
        timeFormat        : 'LST'
    },

    Sidebar : {
        'Filter events' : 'Φιλτράρισμα συμβάντων'
    },

    WeekExpander : {
        expandTip   : 'Κάντε κλικ για να επεκτείνετε τη σειρά',
        collapseTip : 'Κάντε κλικ για να καταργήσετε τη σειρά'
    }
};

export default LocaleHelper.publishLocale(locale);
