// @bryntum/calendar/source/lib/Calendar/localization/Es.js
import LocaleHelper from '../../Core/localization/LocaleHelper.js';
import '../../Scheduler/localization/Es.js';

const locale = {

    localeName : 'Es',
    localeDesc : 'Español',
    localeCode : 'es',

    EventEdit : {
        Calendar  : 'Calendario',
        'All day' : 'Todo el día',
        day       : 'Día',
        week      : 'Semana',
        month     : 'Mes',
        year      : 'Año',
        decade    : 'Década'
    },

    EventMenu : {
        duplicateEvent : 'Duplicar evento',
        copy           : 'copiar'
    },

    Calendar : {
        Today      : 'Hoy',
        next       : range => `Siguiente ${range}`,
        previous   : range => `Anterior ${range}`,
        plusMore   : value => `+${value} más`,
        allDay     : 'Todo el día',
        endsOn     : d => `Termina el ${d}`,
        weekOfYear : ([y, w]) => `Semana ${w}, ${y}`,
        loadFail   : 'Fallo al cargar los datos del calendario. Contacte con el administrador de su sistema'
    },

    CalendarDrag : {
        holdCtrlForRecurrence : 'Presionar CTRL para un evento recurrente'
    },

    CalendarMixin : {
        eventCount : count => `${count || 'Sin'} evento${(count && count > 1) ? 's' : ''}`
    },

    EventTip : {
        'Edit event' : 'Editar evento',
        timeFormat   : 'LST'
    },

    ModeSelector : {
        includeWeekends : 'Incluir fines de semana',
        weekends        : 'Fines de semana'
    },

    AgendaView : {
        Agenda : 'Agenda'
    },

    MonthView : {
        Month     : 'Mes',
        monthUnit : 'mes'
    },

    WeekView : {
        weekUnit : 'semana'
    },

    YearView : {
        Year     : 'Año',
        yearUnit : 'año'
    },

    EventList : {
        List   : 'Lista',
        Start  : 'Inicio',
        Finish : 'Finalización'
    },

    DayView : {
        Day               : 'Día',
        dayUnit           : 'día',
        daysUnit          : 'días',
        expandAllDayRow   : 'Expandir sección de todo el día',
        collapseAllDayRow : 'Contraer sección de todo el día',
        timeFormat        : 'LST'
    },

    Sidebar : {
        'Filter events' : 'Filtrar eventos'
    },

    WeekExpander : {
        expandTip   : 'Hacer clic para expandir fila',
        collapseTip : 'Hacer clic para contraer fila'
    }
};

export default LocaleHelper.publishLocale(locale);
