// @bryntum/calendar/source/lib/Calendar/localization/Et.js
import LocaleHelper from '../../Core/localization/LocaleHelper.js';
import '../../Scheduler/localization/Et.js';

const locale = {

    localeName : 'Et',
    localeDesc : 'Eesti keel',
    localeCode : 'et',

    EventEdit : {
        Calendar  : 'Kalender',
        'All day' : 'Kogu päev',
        day       : 'Päev',
        week      : 'Nädal',
        month     : 'Kuu',
        year      : 'Aasta',
        decade    : 'Kümnend'
    },

    EventMenu : {
        duplicateEvent : 'Dubleeri sündmus',
        copy           : 'kopeeri'
    },

    Calendar : {
        Today      : 'Täna',
        next       : range => `Järgmine ${range}`,
        previous   : range => `Eelmine ${range}`,
        plusMore   : value => `+${value} veel`,
        allDay     : 'Kogu päev',
        endsOn     : d => `Lõpeb ${d}`,
        weekOfYear : ([y, w]) => `Nädal ${w}, ${y}`,
        loadFail   : 'Kalendriandmete laadimine nurjus. Palun võtke ühendust oma süsteemiadministraatoriga'
    },

    CalendarDrag : {
        holdCtrlForRecurrence : 'Korduva sündmuse korral hoidke all klahvi CTRL'
    },

    CalendarMixin : {
        eventCount : count => `${count || 'Ei'} sündmust${(count && count > 1) ? 's' : ''}`
    },

    EventTip : {
        'Edit event' : 'Redigeeri sündmust',
        timeFormat   : 'LST'
    },

    ModeSelector : {
        includeWeekends : 'Kaasa nädalavahetused',
        weekends        : 'Nädalavahetused'
    },

    AgendaView : {
        Agenda : 'Tegevuskava'
    },

    MonthView : {
        Month     : 'Kuu',
        monthUnit : 'kuu'
    },

    WeekView : {
        weekUnit : 'nädal'
    },

    YearView : {
        Year     : 'Aasta',
        yearUnit : 'aasta'
    },

    EventList : {
        List   : 'Loend',
        Start  : 'Algus',
        Finish : 'Lõpp'
    },

    DayView : {
        Day               : 'Päev',
        dayUnit           : 'päev',
        daysUnit          : 'päeva',
        expandAllDayRow   : 'Laienda kogu päeva jaotist',
        collapseAllDayRow : 'Koonda kogu päeva jaotis',
        timeFormat        : 'LST'
    },

    Sidebar : {
        'Filter events' : 'Filtreeri sündmusi'
    },

    WeekExpander : {
        expandTip   : 'Klõpsa rea laiendamiseks',
        collapseTip : 'Klõpsa rea koondamiseks'
    }
};

export default LocaleHelper.publishLocale(locale);
