// @bryntum/calendar/source/lib/Calendar/localization/He.js
import LocaleHelper from '../../Core/localization/LocaleHelper.js';
import '../../Scheduler/localization/He.js';

const locale = {

    localeName : 'He',
    localeDesc : 'עִברִית',
    localeCode : 'he',

    EventEdit : {
        Calendar  : 'לוח שנה',
        'All day' : 'כל היום',
        day       : 'יום',
        week      : 'שבוע',
        month     : 'חודש',
        year      : 'שנה',
        decade    : 'עשור'
    },

    EventMenu : {
        duplicateEvent : 'הכפלת אירוע',
        copy           : 'העתק'
    },

    Calendar : {
        Today      : 'היום',
        next       : range => `${range} הבא`,
        previous   : range => `${range} קודם`,
        plusMore   : value => `+${value} עוד`,
        allDay     : 'כל היום',
        endsOn     : d => `${d} מסתיים`,
        weekOfYear : ([y, w]) => `${y}, ${w}, שבוע`,
        loadFail   : 'טעינת נתוני לוח השנה כשלה. אנא פנה למנהל המערכת שלך.'
    },

    CalendarDrag : {
        holdCtrlForRecurrence : 'לחץ לחיצה ממושכת על מקש קונטרול לאירוע חוזר'
    },

    CalendarMixin : {
        eventCount : count => `${(count && count > 1) ? 'ים' : ''}אירוע ${count || 'אין'}`
    },

    EventTip : {
        'Edit event' : 'ערוך אירוע',
        timeFormat   : 'LST'
    },

    ModeSelector : {
        includeWeekends : 'כלול סופי שבוע',
        weekends        : 'סופי שבוע'
    },

    AgendaView : {
        Agenda : 'יומן'
    },

    MonthView : {
        Month     : 'חודש',
        monthUnit : 'חודש'
    },

    WeekView : {
        weekUnit : 'שבוע'
    },

    YearView : {
        Year     : 'שנה',
        yearUnit : 'שנה'
    },

    EventList : {
        List   : 'רשימה',
        Start  : 'התחל',
        Finish : 'סיים'
    },

    DayView : {
        Day               : 'יום',
        dayUnit           : 'יום',
        daysUnit          : 'ימים',
        expandAllDayRow   : 'הרחב את מקטע כל הימים',
        collapseAllDayRow : 'מזער את מקטע כל הימים',
        timeFormat        : 'LST'
    },

    Sidebar : {
        'Filter events' : 'סנן אירועים'
    },

    WeekExpander : {
        expandTip   : 'הקלק להרחבת השורה',
        collapseTip : 'הקלק למזעור השורה'
    }
};

export default LocaleHelper.publishLocale(locale);
