// @bryntum/calendar/source/lib/Calendar/localization/Hr.js
import LocaleHelper from '../../Core/localization/LocaleHelper.js';
import '../../Scheduler/localization/Hr.js';

const locale = {

    localeName : 'Hr',
    localeDesc : 'Hrvatski',
    localeCode : 'hr',

    EventEdit : {
        Calendar  : 'Kalendar',
        'All day' : 'Cijeli dan',
        day       : 'Dan',
        week      : 'Tjedan',
        month     : 'Mjesec',
        year      : 'Godina',
        decade    : 'Desetljeće'
    },

    EventMenu : {
        duplicateEvent : 'Dupliciraj događaj',
        copy           : 'kopiraj'
    },

    Calendar : {
        Today      : 'Danas',
        next       : range => `Sljedeći ${range}`,
        previous   : range => `Prethodni ${range}`,
        plusMore   : value => `+${value} više`,
        allDay     : 'Cijeli dan',
        endsOn     : d => `Završava ${d}`,
        weekOfYear : ([y, w]) => `Tjedan ${w}, ${y}`,
        loadFail   : 'Učitavanje podataka kalendara nije uspjelo. Molimo kontaktirajte svog administratora sustava'
    },

    CalendarDrag : {
        holdCtrlForRecurrence : 'Drži CTRL za ponavljajući događaj'
    },

    CalendarMixin : {
        eventCount : count => `${count || 'Nema'} događaj${(count && count > 1) ? 'a' : '' }`
    },

    EventTip : {
        'Edit event' : 'Uredi događaj',
        timeFormat   : 'LST'
    },

    ModeSelector : {
        includeWeekends : 'Uključujući vikende',
        weekends        : 'Vikendi'
    },

    AgendaView : {
        Agenda : 'Raspored događanja'
    },

    MonthView : {
        Month     : 'Mjesec',
        monthUnit : 'mjesec'
    },

    WeekView : {
        weekUnit : 'tjedan'
    },

    YearView : {
        Year     : 'Godina',
        yearUnit : 'godina'
    },

    EventList : {
        List   : 'Popis',
        Start  : 'Početak',
        Finish : 'Završetak'
    },

    DayView : {
        Day               : 'Dan',
        dayUnit           : 'dan',
        daysUnit          : 'dani',
        expandAllDayRow   : 'Proširi odjeljak cijeloga dana',
        collapseAllDayRow : 'Sažmi odjeljak cijeloga dana',
        timeFormat        : 'LST'
    },

    Sidebar : {
        'Filter events' : 'Filtriraj događaje'
    },

    WeekExpander : {
        expandTip   : 'Klikni za proširivanje retka',
        collapseTip : 'Klikni za sažimanje retka'
    }
};

export default LocaleHelper.publishLocale(locale);
