// @bryntum/calendar/source/lib/Calendar/localization/Hu.js
import LocaleHelper from '../../Core/localization/LocaleHelper.js';
import '../../Scheduler/localization/Hu.js';

const locale = {

    localeName : 'Hu',
    localeDesc : 'Magyar',
    localeCode : 'hu',

    EventEdit : {
        Calendar  : 'Naptár',
        'All day' : 'Egész nap',
        day       : 'Nap',
        week      : 'Hét',
        month     : 'Hónap',
        year      : 'Év',
        decade    : 'Évtized'
    },

    EventMenu : {
        duplicateEvent : 'Esemény duplikálása',
        copy           : 'másolás'
    },

    Calendar : {
        Today      : 'Másolás',
        next       : range => `Következő ${range}`,
        previous   : range => `Előző ${range}`,
        plusMore   : value => `+${value} további`,
        allDay     : 'Egész nap',
        endsOn     : d => `Vége ${d}`,
        weekOfYear : ([y, w]) => `Hét ${w}, ${y}`,
        loadFail   : 'A naptáradatok betöltése sikertelen. Forduljon a rendszergazdához.'
    },

    CalendarDrag : {
        holdCtrlForRecurrence : 'Ismétlődő eseményhez nyomja a CTRL-t'
    },

    CalendarMixin : {
        eventCount : count => `${count || 'Nincs'} esemény`
    },

    EventTip : {
        'Edit event' : 'Esemény szerkesztése',
        timeFormat   : 'LST'
    },

    ModeSelector : {
        includeWeekends : 'Hétvégék beszámítása',
        weekends        : 'Hétvégék'
    },

    AgendaView : {
        Agenda : 'Napirend'
    },

    MonthView : {
        Month     : 'Hónap',
        monthUnit : 'hónap'
    },

    WeekView : {
        weekUnit : 'hét'
    },

    YearView : {
        Year     : 'Év',
        yearUnit : 'év'
    },

    EventList : {
        List   : 'Lista',
        Start  : 'Kezdés',
        Finish : 'Vége'
    },

    DayView : {
        Day               : 'Nap',
        dayUnit           : 'nap',
        daysUnit          : 'nap',
        expandAllDayRow   : 'Egész nap rész kibontása',
        collapseAllDayRow : 'Egész nap rész összecsukása',
        timeFormat        : 'LST'
    },

    Sidebar : {
        'Filter events' : 'Események szűrése'
    },

    WeekExpander : {
        expandTip   : 'Kattintson a sor kibontásához',
        collapseTip : 'Kattintson a sor összecsukásához'
    }
};

export default LocaleHelper.publishLocale(locale);
