// @bryntum/calendar/source/lib/Calendar/localization/Id.js
import LocaleHelper from '../../Core/localization/LocaleHelper.js';
import '../../Scheduler/localization/Id.js';

const locale = {

    localeName : 'Id',
    localeDesc : 'Bahasa Indonesia',
    localeCode : 'id',

    EventEdit : {
        Calendar  : 'Kalender',
        'All day' : 'Sepanjang hari',
        day       : 'Hari',
        week      : 'Minggu',
        month     : 'Bulan',
        year      : 'Tahun',
        decade    : 'Dekade'
    },

    EventMenu : {
        duplicateEvent : 'Duplikasi acara',
        copy           : 'salin'
    },

    Calendar : {
        Today      : 'Hari ini',
        next       : range => `${range} Berikutnya`,
        previous   : range => `${range} Sebelumnya`,
        plusMore   : value => `+${value} lagi`,
        allDay     : 'Sepanjang hari',
        endsOn     : d => `Berakhir ${d}`,
        weekOfYear : ([y, w]) => `Minggu ${w}, ${y}`,
        loadFail   : 'Pemuatan data kalender gagal. Harap hubungi administrator sistem Anda'
    },

    CalendarDrag : {
        holdCtrlForRecurrence : 'Tahan CTRL untuk acara rutin'
    },

    CalendarMixin : {
        eventCount : count => `${count || 'Tidak ada'} event`
    },

    EventTip : {
        'Edit event' : 'Edit acara',
        timeFormat   : 'LST'
    },

    ModeSelector : {
        includeWeekends : 'Termasuk akhir pekan',
        weekends        : 'Akhir pekan'
    },

    AgendaView : {
        Agenda : 'Agenda'
    },

    MonthView : {
        Month     : 'Bulan',
        monthUnit : 'bulan'
    },

    WeekView : {
        weekUnit : 'minggu'
    },

    YearView : {
        Year     : 'Tahun',
        yearUnit : 'tahun'
    },

    EventList : {
        List   : 'Daftar',
        Start  : 'Mulai',
        Finish : 'Selesai'
    },

    DayView : {
        Day               : 'Hari',
        dayUnit           : 'hari',
        daysUnit          : 'hari',
        expandAllDayRow   : 'Luaskan bagian sepanjang hari',
        collapseAllDayRow : 'Ciutkan bagian sepanjang hari',
        timeFormat        : 'LST'
    },

    Sidebar : {
        'Filter events' : 'Filter acara'
    },

    WeekExpander : {
        expandTip   : 'Klik untuk memperluas baris',
        collapseTip : 'Klik untuk menciutkan baris'
    }
};

export default LocaleHelper.publishLocale(locale);
