// @bryntum/calendar/source/lib/Calendar/localization/It.js
import LocaleHelper from '../../Core/localization/LocaleHelper.js';
import '../../Scheduler/localization/It.js';

const locale = {

    localeName : 'It',
    localeDesc : 'Italiano',
    localeCode : 'it',

    EventEdit : {
        Calendar  : 'Calendario',
        'All day' : 'Tutto il giorno',
        day       : 'Giorno',
        week      : 'Settimana',
        month     : 'Mese',
        year      : 'Anno',
        decade    : 'Decennio'
    },

    EventMenu : {
        duplicateEvent : 'Duplica evento',
        copy           : 'copia'
    },

    Calendar : {
        Today      : 'Oggi',
        next       : range => `Prossimo ${range}`,
        previous   : range => `Scorso ${range}`,
        plusMore   : value => `altri +${value}`,
        allDay     : 'Tutto il giorno',
        endsOn     : d => `Termina ${d}`,
        weekOfYear : ([y, w]) => `Settimana ${w}, ${y}`,
        loadFail   : 'Caricamento dei dati del calendario non riuscito. Contatta il tuo amministratore di sistema'
    },

    CalendarDrag : {
        holdCtrlForRecurrence : 'Tieni premuto CTRL per un evento ricorrente'
    },

    CalendarMixin : {
        eventCount : count => `${count || 'Nessun'} event${(count && count > 1) ? 'i' : 'o'}`
    },

    EventTip : {
        'Edit event' : 'Modifica evento',
        timeFormat   : 'LST'
    },

    ModeSelector : {
        includeWeekends : 'Includi fine settimana',
        weekends        : 'Fine settimana'
    },

    AgendaView : {
        Agenda : 'Agenda'
    },

    MonthView : {
        Month     : 'Mese',
        monthUnit : 'mese'
    },

    WeekView : {
        weekUnit : 'settimana'
    },

    YearView : {
        Year     : 'Anno',
        yearUnit : 'anno'
    },

    EventList : {
        List   : 'Elenco',
        Start  : 'Inizio',
        Finish : 'Fine'
    },

    DayView : {
        Day               : 'Giorno',
        dayUnit           : 'giorno',
        daysUnit          : 'giorni',
        expandAllDayRow   : 'Espandi la sezione di tutti i giorni',
        collapseAllDayRow : 'Comprimi la sezione di tutti i giorni',
        timeFormat        : 'LST'
    },

    Sidebar : {
        'Filter events' : 'Filtra eventi'
    },

    WeekExpander : {
        expandTip   : 'Fai clic per espandere la riga',
        collapseTip : 'Fai clic per comprimere la riga'
    }
};

export default LocaleHelper.publishLocale(locale);
