// @bryntum/calendar/source/lib/Calendar/localization/Ja.js
import LocaleHelper from '../../Core/localization/LocaleHelper.js';
import '../../Scheduler/localization/Ja.js';

const locale = {

    localeName : 'Ja',
    localeDesc : '日本語',
    localeCode : 'ja',

    EventEdit : {
        Calendar  : '予定表',
        'All day' : '終日',
        day       : '日',
        week      : '週',
        month     : '月',
        year      : '年',
        decade    : '十年'
    },

    EventMenu : {
        duplicateEvent : 'イベントを繰り返す',
        copy           : 'コピーする'
    },

    Calendar : {
        Today      : '今日',
        next       : range => `次へ ${range}`,
        previous   : range => `前へ ${range}`,
        plusMore   : value => `+${value} 増`,
        allDay     : '終日',
        endsOn     : d => `終了 ${d}`,
        weekOfYear : ([y, w]) => `週 ${w}, ${y}`,
        loadFail   : 'カレンダーデータの読み込みに失敗しました。システム管理者に連絡してください'
    },

    CalendarDrag : {
        holdCtrlForRecurrence : '繰り返しイベントはCTRLキーを押したままにする'
    },

    CalendarMixin : {
        eventCount : count => `${count || 'なし'}ベント$`
    },

    EventTip : {
        'Edit event' : 'イベントを編集する',
        timeFormat   : 'LST'
    },

    ModeSelector : {
        includeWeekends : '週末を含む',
        weekends        : '週末'
    },

    AgendaView : {
        Agenda : 'アジェンダ'
    },

    MonthView : {
        Month     : '月',
        monthUnit : '月'
    },

    WeekView : {
        weekUnit : '週'
    },

    YearView : {
        Year     : '年',
        yearUnit : '年'
    },

    EventList : {
        List   : 'リスト',
        Start  : '開始',
        Finish : '終了'
    },

    DayView : {
        Day               : '日',
        dayUnit           : '日',
        daysUnit          : '日',
        expandAllDayRow   : '終日セクションを拡大する',
        collapseAllDayRow : '終日セクションを縮小する',
        timeFormat        : 'LST'
    },

    Sidebar : {
        'Filter events' : 'イベントをフィルターする'
    },

    WeekExpander : {
        expandTip   : 'クリックして行を拡大する',
        collapseTip : 'クリックして行を縮小する'
    }
};

export default LocaleHelper.publishLocale(locale);
