// @bryntum/calendar/source/lib/Calendar/localization/Ms.js
import LocaleHelper from '../../Core/localization/LocaleHelper.js';
import '../../Scheduler/localization/Ms.js';

const locale = {

    localeName : 'Ms',
    localeDesc : 'Melayu',
    localeCode : 'ms',

    EventEdit : {
        Calendar  : 'Kalendar',
        'All day' : 'Sepanjang hari',
        day       : 'Hari',
        week      : 'Minggu',
        month     : 'Bulan',
        year      : 'Tahun',
        decade    : 'Dekad'
    },

    EventMenu : {
        duplicateEvent : 'Duplikat peristiwa',
        copy           : 'salin'
    },

    Calendar : {
        Today      : 'Hari ini',
        next       : range => `Berikutnya ${range}`,
        previous   : range => `Sebelumnya ${range}`,
        plusMore   : value => `+${value} lagi`,
        allDay     : 'Sepanjang hari',
        endsOn     : d => `Berakhir ${d}`,
        weekOfYear : ([y, w]) => `Minggu ${w}, ${y}`,
        loadFail   : 'Pemuatan data kalendar gagal. Sila hubungi pentadbir sistem anda'
    },

    CalendarDrag : {
        holdCtrlForRecurrence : 'Tahan CTRL untuk peristiwa berulang'
    },

    CalendarMixin : {
        eventCount : count => `${count || 'Jumlah'} peristiwa`
    },

    EventTip : {
        'Edit event' : 'Ubah peristiwa',
        timeFormat   : 'LST'
    },

    ModeSelector : {
        includeWeekends : 'Termasuk hujung minggu',
        weekends        : 'Hujung minggu'
    },

    AgendaView : {
        Agenda : 'Agenda'
    },

    MonthView : {
        Month     : 'Bulan',
        monthUnit : 'bulan'
    },

    WeekView : {
        weekUnit : 'minggu'
    },

    YearView : {
        Year     : 'Tahun',
        yearUnit : 'tahun'
    },

    EventList : {
        List   : 'Senarai',
        Start  : 'Mula',
        Finish : 'Selesai'
    },

    DayView : {
        Day               : 'Hari',
        dayUnit           : 'hari',
        daysUnit          : 'hari',
        expandAllDayRow   : 'Kembangkan bahagian sepanjang hari',
        collapseAllDayRow : 'Kecilkan bahagian sepanjang hari',
        timeFormat        : 'LST'
    },

    Sidebar : {
        'Filter events' : 'Tapis peristiwa'
    },

    WeekExpander : {
        expandTip   : 'Klik untuk kembangkan baris',
        collapseTip : 'Klik untuk kecilkan baris'
    }
};

export default LocaleHelper.publishLocale(locale);
