// @bryntum/calendar/source/lib/Calendar/localization/Nl.js
import LocaleHelper from '../../Core/localization/LocaleHelper.js';
import '../../Scheduler/localization/Nl.js';

const locale = {

    localeName : 'Nl',
    localeDesc : 'Nederlands',
    localeCode : 'nl',

    EventEdit : {
        Calendar  : 'Kalender',
        'All day' : 'Hele dag',
        day       : 'Dag',
        week      : 'Week',
        month     : 'Maand',
        year      : 'Jaar',
        decade    : 'Decennium'
    },

    EventMenu : {
        duplicateEvent : 'Kopiëren boeking',
        copy           : 'kopie'
    },

    Calendar : {
        Today      : 'Vandaag',
        next       : range => `Volgende ${range}`,
        previous   : range => `Vorige ${range}`,
        plusMore   : value => `+${value} meer`,
        allDay     : 'Hele dag',
        endsOn     : d => `Eindigt op ${d}`,
        weekOfYear : ([y, w]) => `Week ${w}, ${y}`,
        loadFail   : 'Het laden van agendagegevens is mislukt. Neem contact op met uw systeembeheerder'
    },

    CalendarDrag : {
        holdCtrlForRecurrence : 'Toets CTRL voor herhalende gebeurtenis'
    },

    CalendarMixin : {
        eventCount : count => `${count || 'No'} event${count === 1 ? '' : 's'}`
    },

    EventTip : {
        'Edit event' : 'Bewerk boeking',
        timeFormat   : 'LT'
    },

    ModeSelector : {
        includeWeekends : 'Inkluder helger',
        weekends        : 'Helger'
    },

    AgendaView : {
        Agenda : 'Agenda'
    },

    MonthView : {
        Month     : 'Maand',
        monthUnit : 'maand'
    },

    WeekView : {
        weekUnit : 'week'
    },

    YearView : {
        Year     : 'Jaar',
        yearUnit : 'jaar'
    },

    EventList : {
        List   : 'Lijst',
        Start  : 'Begin',
        Finish : 'Af hebben'
    },

    DayView : {
        Day               : 'Dag',
        dayUnit           : 'dag',
        daysUnit          : 'dagen',
        expandAllDayRow   : 'Klap dag overzicht uit',
        collapseAllDayRow : 'Klap dag overzicht in',
        timeFormat        : 'LT'
    },

    Sidebar : {
        'Filter events' : 'Filter kalender items'
    },

    WeekExpander : {
        expandTip   : 'Klik om de rij uit te vouwen',
        collapseTip : 'Klik om de rij samen te vouwen'
    }
};

export default LocaleHelper.publishLocale(locale);
