// @bryntum/calendar/source/lib/Calendar/localization/Pt.js
import LocaleHelper from '../../Core/localization/LocaleHelper.js';
import '../../Scheduler/localization/Pt.js';

const locale = {

    localeName : 'Pt',
    localeDesc : 'Português',
    localeCode : 'pt',

    EventEdit : {
        Calendar  : 'Calendário',
        'All day' : 'Todo o dia',
        day       : 'Dia',
        week      : 'Semana',
        month     : 'Mês',
        year      : 'Ano',
        decade    : 'Década'
    },

    EventMenu : {
        duplicateEvent : 'Duplicar evento',
        copy           : 'copiar'
    },

    Calendar : {
        Today      : 'Hoje',
        next       : range => `Seguinte ${range}`,
        previous   : range => `Anterior ${range}`,
        plusMore   : value => `+${value} mais`,
        allDay     : 'Todo o dia',
        endsOn     : d => `Acaba em ${d}`,
        weekOfYear : ([y, w]) => `Semana ${w}, ${y}`,
        loadFail   : 'Falha ao carregar dados do calendário. Contacte o administrador do sistema'
    },

    CalendarDrag : {
        holdCtrlForRecurrence : 'Prima continuamente CTRL para um evento recorrente'
    },

    CalendarMixin : {
        eventCount : count => `${count || 'Sin'} evento${(count && count > 1) ? '' : 's'}`
    },

    EventTip : {
        'Edit event' : 'Editar evento',
        timeFormat   : 'LST'
    },

    ModeSelector : {
        includeWeekends : 'Incluir fins-de-semana',
        weekends        : 'Fins-de-semana'
    },

    AgendaView : {
        Agenda : 'Agenda'
    },

    MonthView : {
        Month     : 'Mês',
        monthUnit : 'mês'
    },

    WeekView : {
        weekUnit : 'semana'
    },

    YearView : {
        Year     : 'Ano',
        yearUnit : 'ano'
    },

    EventList : {
        List   : 'Lista',
        Start  : 'Início',
        Finish : 'Fim'
    },

    DayView : {
        Day               : 'Dia',
        dayUnit           : 'dia',
        daysUnit          : 'dias',
        expandAllDayRow   : 'Expandir a secção de todo o dia',
        collapseAllDayRow : 'Fechar a secção de todo o dia',
        timeFormat        : 'LST'
    },

    Sidebar : {
        'Filter events' : 'Filtrar eventos'
    },

    WeekExpander : {
        expandTip   : 'Clique para expandir a fila',
        collapseTip : 'Clique para fechar a fila'
    }
};

export default LocaleHelper.publishLocale(locale);
