// @bryntum/calendar/source/lib/Calendar/localization/Ru.js
import LocaleHelper from '../../Core/localization/LocaleHelper.js';
import '../../Scheduler/localization/Ru.js';

const locale = {

    localeName : 'Ru',
    localeDesc : 'Русский',
    localeCode : 'ru',

    EventEdit : {
        Calendar  : 'Календарь',
        'All day' : 'На весь день',
        day       : 'День',
        week      : 'Неделя',
        month     : 'Месяц',
        year      : 'Год',
        decade    : 'Десятилетие'
    },

    EventMenu : {
        duplicateEvent : 'копировать событие',
        copy           : 'копия'
    },

    Calendar : {
        Today : 'Сегодня',
        next  : range => {
            const next = range.includes('неделя') ? 'следующая' : 'следующий';

            return `${next} ${range}`;
        },
        previous : range => {
            const previous = range.includes('неделя') ?  'предыдущая' : 'предыдущий';

            return `${previous} ${range}`;
        },
        plusMore   : value => `Еще +${value}`,
        allDay     : 'На весь день',
        endsOn     : d => `Заканчивается ${d}`,
        weekOfYear : ([y, w]) => `Неделя ${w}, ${y}`,
        loadFail   : 'Не удалось загрузить данные календаря. Пожалуйста, обратитесь к системному администратору'
    },

    CalendarDrag : {
        holdCtrlForRecurrence : 'Удерживайте CTRL для повторяющегося события'
    },

    CalendarMixin : {
        eventCount : count => `${count || 'Нет'} ${count === 1 ? 'событие' : 'событий'}`
    },

    EventTip : {
        'Edit event' : 'Редактировать событие',
        timeFormat   : 'LT'
    },

    ModeSelector : {
        includeWeekends : 'Включить выходные',
        weekends        : 'выходные'
    },

    AgendaView : {
        Agenda : 'Повестка дня'
    },

    MonthView : {
        Month     : 'Месяц',
        monthUnit : 'месяц'
    },

    WeekView : {
        weekUnit : 'неделя'
    },

    YearView : {
        Year     : 'Год',
        yearUnit : 'год'
    },

    EventList : {
        List   : 'Список',
        Start  : 'Начало',
        Finish : 'Конец'
    },

    DayView : {
        Day               : 'День',
        dayUnit           : 'день',
        daysUnit          : 'дни',
        expandAllDayRow   : 'Развернуть группу событий на весь день',
        collapseAllDayRow : 'Свернуть группу событий на весь день',
        timeFormat        : 'LT'
    },

    Sidebar : {
        'Filter events' : 'Фильтровать события'
    },

    WeekExpander : {
        expandTip   : 'Нажмите, чтобы развернуть строку',
        collapseTip : 'Нажмите, чтобы свернуть строку'
    }
};

export default LocaleHelper.publishLocale(locale);
