// @bryntum/calendar/source/lib/Calendar/localization/Sk.js
import LocaleHelper from '../../Core/localization/LocaleHelper.js';
import '../../Scheduler/localization/Sk.js';

const locale = {

    localeName : 'Sk',
    localeDesc : 'Slovenský',
    localeCode : 'sk',

    EventEdit : {
        Calendar  : 'Kalendár',
        'All day' : 'Celý deň',
        day       : 'Deň',
        week      : 'Týždeň',
        month     : 'Mesiac',
        year      : 'Rok',
        decade    : 'Dekáda'
    },

    EventMenu : {
        duplicateEvent : 'Duplicitná udalosť',
        copy           : 'kopírovať'
    },

    Calendar : {
        Today      : 'Dnes',
        next       : range => `Ďalší ${range}`,
        previous   : range => `Predchádzajúci ${range}`,
        plusMore   : value => `+${value} viac`,
        allDay     : 'Celý deň',
        endsOn     : d => `Končí ${d}`,
        weekOfYear : ([y, w]) => `Týždeň ${w}, ${y}`,
        loadFail   : 'Načítanie údajov z kalendára zlyhalo. Kontaktujte správcu systému'
    },

    CalendarDrag : {
        holdCtrlForRecurrence : 'Podržte CTRL pre opakujúcu sa udalosť'
    },

    CalendarMixin : {
        eventCount : count => `${count || 'Žiadne'} udalos${(count && count > 1) ? 'ti' : 'ť' }`
    },

    EventTip : {
        'Edit event' : 'Upraviť udalosť',
        timeFormat   : 'LST'
    },

    ModeSelector : {
        includeWeekends : 'Zahrnúť víkendy',
        weekends        : 'Víkendy'
    },

    AgendaView : {
        Agenda : 'Agenda'
    },

    MonthView : {
        Month     : 'Mesiac',
        monthUnit : 'mesiac'
    },

    WeekView : {
        weekUnit : 'týždeň'
    },

    YearView : {
        Year     : 'Rok',
        yearUnit : 'rok'
    },

    EventList : {
        List   : 'Zoznam',
        Start  : 'Začať',
        Finish : 'Skončiť'
    },

    DayView : {
        Day               : 'Deň',
        dayUnit           : 'deň',
        daysUnit          : 'dni',
        expandAllDayRow   : 'Rozšíriť časť celého dňa',
        collapseAllDayRow : 'Zložiť časť celého dňa',
        timeFormat        : 'LST'
    },

    Sidebar : {
        'Filter events' : 'Filtrovať udalosti'
    },

    WeekExpander : {
        expandTip   : 'Kliknúť pre rozšírenie riadka',
        collapseTip : 'Kliknúť pre zloženie riadka'
    }
};

export default LocaleHelper.publishLocale(locale);
