// @bryntum/calendar/source/lib/Calendar/localization/Sl.js
import LocaleHelper from '../../Core/localization/LocaleHelper.js';
import '../../Scheduler/localization/Sl.js';

const locale = {

    localeName : 'Sl',
    localeDesc : 'Slovensko',
    localeCode : 'sl',

    EventEdit : {
        Calendar  : 'Koledar',
        'All day' : 'Ves dan',
        day       : 'Dan',
        week      : 'Teden',
        month     : 'Mesec',
        year      : 'Leto',
        decade    : 'Desetletje'
    },

    EventMenu : {
        duplicateEvent : 'Podvojeni dogodek',
        copy           : 'kopiraj'
    },

    Calendar : {
        Today      : 'Danes',
        next       : range => `Naslednji ${range}`,
        previous   : range => `Prejšnji ${range}`,
        plusMore   : value => `+${value} več`,
        allDay     : 'Ves dan',
        endsOn     : d => `Konec ${d}`,
        weekOfYear : ([y, w]) => `Teden ${w}, ${y}`,
        loadFail   : 'Nalaganje podatkov iz koledarja ni uspelo. Obrnite se na skrbnika sistema'
    },

    CalendarDrag : {
        holdCtrlForRecurrence : 'Drži CTRL za ponavljajoči se dogodek'
    },

    CalendarMixin : {
        eventCount : count => `${count || 'Brez'} dogod${(count && count > 1) ? 'ka' : 'ek' }`
    },

    EventTip : {
        'Edit event' : 'Uredi dogodek',
        timeFormat   : 'LST'
    },

    ModeSelector : {
        includeWeekends : 'Vključno z vikendi',
        weekends        : 'Vikendi'
    },

    AgendaView : {
        Agenda : 'Dnevni red'
    },

    MonthView : {
        Month     : 'Mesec',
        monthUnit : 'mesec'
    },

    WeekView : {
        weekUnit : 'Teden'
    },

    YearView : {
        Year     : 'Leto',
        yearUnit : 'Leto'
    },

    EventList : {
        List   : 'Seznam',
        Start  : 'Začetek',
        Finish : 'Konec'
    },

    DayView : {
        Day               : 'Dan',
        dayUnit           : 'dan',
        daysUnit          : 'dnevi',
        expandAllDayRow   : 'Razširi celodnevni razdelek',
        collapseAllDayRow : 'Strni celodnevni razdelek',
        timeFormat        : 'LST'
    },

    Sidebar : {
        'Filter events' : 'Filtriraj dogodke'
    },

    WeekExpander : {
        expandTip   : 'Klikni za razširitev vrstice',
        collapseTip : 'Klikni za strnitev vrstice'
    }
};

export default LocaleHelper.publishLocale(locale);
