// @bryntum/calendar/source/lib/Calendar/localization/Sr.js
import LocaleHelper from '../../Core/localization/LocaleHelper.js';
import '../../Scheduler/localization/Sr.js';

const locale = {

    localeName : 'Sr',
    localeDesc : 'Srpski',
    localeCode : 'sr',

    EventEdit : {
        Calendar  : 'Kalendar',
        'All day' : 'Ceo dan',
        day       : 'Dan',
        week      : 'Nedelja',
        month     : 'Mesec',
        year      : 'Godina',
        decade    : 'Decenija'
    },

    EventMenu : {
        duplicateEvent : 'Dupliraj događaj',
        copy           : 'kopiraj'
    },

    Calendar : {
        Today      : 'Danas',
        next       : range => `Sledeći ${range}`,
        previous   : range => `Prethodni ${range}`,
        plusMore   : value => `+${value} više`,
        allDay     : 'Ceo dan',
        endsOn     : d => `Završava ${d}`,
        weekOfYear : ([y, w]) => `Nedelja ${w}, ${y}`,
        loadFail   : 'Učitavanje podataka kalendara nije uspelo. Obratite se svom sistemskom administratoru'
    },

    CalendarDrag : {
        holdCtrlForRecurrence : 'Držite CTRL za događaje koji se ponavljaju'
    },

    CalendarMixin : {
        eventCount : count => `${count || 'Nema'} događaj${(count && count > 1) ? 'e' : '' }`
    },

    EventTip : {
        'Edit event' : 'Uredi događaj',
        timeFormat   : 'LST'
    },

    ModeSelector : {
        includeWeekends : 'Uključi vikende',
        weekends        : 'Vikendi'
    },

    AgendaView : {
        Agenda : 'Rokovnik'
    },

    MonthView : {
        Month     : 'Mesec',
        monthUnit : 'mesec'
    },

    WeekView : {
        weekUnit : 'nedelja'
    },

    YearView : {
        Year     : 'Godina',
        yearUnit : 'godina'
    },

    EventList : {
        List   : 'Lista',
        Start  : 'Početak',
        Finish : 'Kraj'
    },

    DayView : {
        Day               : 'Dan',
        dayUnit           : 'dan',
        daysUnit          : 'dana',
        expandAllDayRow   : 'Proširi red ceo dan',
        collapseAllDayRow : 'Skupi red ceo dan',
        timeFormat        : 'LST'
    },

    Sidebar : {
        'Filter events' : 'Filtriraj događaje'
    },

    WeekExpander : {
        expandTip   : 'Klkni da raširiš red',
        collapseTip : 'Klikni da skupiš red'
    }
};

export default LocaleHelper.publishLocale(locale);
