// @bryntum/calendar/source/lib/Calendar/localization/SvSE.js
import LocaleHelper from '../../Core/localization/LocaleHelper.js';
import '../../Scheduler/localization/SvSE.js';

const locale = {

    localeName : 'SvSE',
    localeDesc : 'Svenska',
    localeCode : 'sv-SE',

    EventEdit : {
        Calendar  : 'Kalender',
        'All day' : 'Heldag',
        day       : 'Dag',
        week      : 'Vecka',
        month     : 'Månad',
        year      : 'År',
        decade    : 'Årtionde'
    },

    EventMenu : {
        duplicateEvent : 'Kopiera bokning',
        copy           : 'kopia'
    },

    Calendar : {
        Today      : 'Idag',
        next       : range => `Nästa ${range}`,
        previous   : range => `Föregående ${range}`,
        plusMore   : value => `+${value} more`,
        allDay     : 'Heldag',
        endsOn     : d => `Slutar ${d}`,
        weekOfYear : ([y, w]) => `Vecka ${w}, ${y}`,
        loadFail   : 'Det gick inte att ladda kalenderdata. Kontakta din systemadministratör'
    },

    CalendarDrag : {
        holdCtrlForRecurrence : 'Håll ner CTRL för upprepande händelse'
    },

    CalendarMixin : {
        eventCount : count => `${count || 'Inga'} händelse${count === 1 ? '' : 'r'}`
    },

    EventTip : {
        'Edit event' : 'Redigera bokning',
        timeFormat   : 'LT'
    },

    ModeSelector : {
        includeWeekends : 'Inkluder helger',
        weekends        : 'Helger'
    },

    AgendaView : {
        Agenda : 'Dagordning'
    },

    MonthView : {
        Month     : 'Månad',
        monthUnit : 'månad'
    },

    WeekView : {
        weekUnit : 'Vecka'
    },

    YearView : {
        Year     : 'År',
        yearUnit : 'år'
    },

    EventList : {
        List   : 'Lista',
        Start  : 'Start',
        Finish : 'Avsluta'
    },

    DayView : {
        Day               : 'Dag',
        dayUnit           : 'dag',
        daysUnit          : 'dagar',
        expandAllDayRow   : 'Expandera heldagssektionen',
        collapseAllDayRow : 'Fäll ihop heldagssektionen',
        timeFormat        : 'LT'
    },

    Sidebar : {
        'Filter events' : 'Filtrera händelser'
    },

    WeekExpander : {
        expandTip   : 'Klicka för att expandera raden',
        collapseTip : 'Klicka för att dölja radenу'
    }
};

export default LocaleHelper.publishLocale(locale);
