// @bryntum/calendar/source/lib/Calendar/localization/Tr.js
import LocaleHelper from '../../Core/localization/LocaleHelper.js';
import '../../Scheduler/localization/Tr.js';

const locale = {

    localeName : 'Tr',
    localeDesc : 'Türkçe',
    localeCode : 'tr',

    EventEdit : {
        Calendar  : 'Takvim',
        'All day' : 'Tüm gün',
        day       : 'Gün',
        week      : 'Hafta',
        month     : 'Ay',
        year      : 'Yıl',
        decade    : 'On yıl'
    },

    EventMenu : {
        duplicateEvent : 'Yinelenen etkinlik',
        copy           : 'kopyala'
    },

    Calendar : {
        Today      : 'Bugün',
        next       : range => `Sonraki ${range}`,
        previous   : range => `Önceki ${range}`,
        plusMore   : value => `+${value} daha fazla`,
        allDay     : 'Tüm gün',
        endsOn     : d => `Bitiş tarihi ${d}`,
        weekOfYear : ([y, w]) => `Hafta ${w}, ${y}`,
        loadFail   : 'Takvim verisi yüklemesi başarısız. Lütfen sistem yöneticiniz ile iletişime geçin.'
    },

    CalendarDrag : {
        holdCtrlForRecurrence : 'Yinelenen etkinlik için CTRL’ye basılı tutun'
    },

    CalendarMixin : {
        eventCount : count => `${count || 'Yok'} Etkinli${(count && count > 1) ? 'kleri' : 'ği'}`
    },

    EventTip : {
        'Edit event' : 'Etkinliği düzenle',
        timeFormat   : 'LST'
    },

    ModeSelector : {
        includeWeekends : 'Hafta sonlarını dahil et',
        weekends        : 'Hafta sonları'
    },

    AgendaView : {
        Agenda : 'Ajanda'
    },

    MonthView : {
        Month     : 'Ay',
        monthUnit : 'ay'
    },

    WeekView : {
        weekUnit : 'hafta'
    },

    YearView : {
        Year     : 'Yıl',
        yearUnit : 'yıl'
    },

    EventList : {
        List   : 'Liste',
        Start  : 'Başlangıç',
        Finish : 'Son'
    },

    DayView : {
        Day               : 'Gün',
        dayUnit           : 'gün',
        daysUnit          : 'gün',
        expandAllDayRow   : 'Tüm gün alanını genişlet',
        collapseAllDayRow : 'Tüm gün alanını daralt',
        timeFormat        : 'LST'
    },

    Sidebar : {
        'Filter events' : 'Etkinlikleri filtrele'
    },

    WeekExpander : {
        expandTip   : 'Satırı genişletmek için tıkla',
        collapseTip : 'Satırı daraltmak için tıkla'
    }
};

export default LocaleHelper.publishLocale(locale);
