// @bryntum/calendar/source/lib/Calendar/localization/Vi.js
import LocaleHelper from '../../Core/localization/LocaleHelper.js';
import '../../Scheduler/localization/Vi.js';

const locale = {

    localeName : 'Vi',
    localeDesc : 'Tiếng Việt',
    localeCode : 'vi',

    EventEdit : {
        Calendar  : 'Lịch',
        'All day' : 'Cả ngày',
        day       : 'Ngày',
        week      : 'Tuần',
        month     : 'Tháng',
        year      : 'Năm',
        decade    : 'Thập kỷ'
    },

    EventMenu : {
        duplicateEvent : 'Sự kiện lặp lại',
        copy           : 'sao chép'
    },

    Calendar : {
        Today      : 'Hôm nay',
        next       : range => `Sau ${range}`,
        previous   : range => `Trước ${range}`,
        plusMore   : value => `+ thêm ${value} `,
        allDay     : 'Cả ngày',
        endsOn     : d => `Kết thúc vào ngày ${d}`,
        weekOfYear : ([y, w]) => `Tuần trong năm${w}, ${y}`,
        loadFail   : 'Tải dữ liệu lịch không thành công. Vui lòng liên hệ với quản trị hệ thống của quý vị'
    },

    CalendarDrag : {
        holdCtrlForRecurrence : 'Giữ nút CTRL cho một sự kiện lặp lại'
    },

    CalendarMixin : {
        eventCount : count => `${count || 'không'} sự kiện`
    },

    EventTip : {
        'Edit event' : 'Chỉnh sửa sự kiện',
        timeFormat   : 'LST'
    },

    ModeSelector : {
        includeWeekends : 'Bao gồm những ngày cuối tuần',
        weekends        : 'Những ngày cuối tuần'
    },

    AgendaView : {
        Agenda : 'Lịch trình'
    },

    MonthView : {
        Month     : 'Tháng',
        monthUnit : 'Đơn vị tháng'
    },

    WeekView : {
        weekUnit : 'Đơn vị tuần'
    },

    YearView : {
        Year     : 'Năm',
        yearUnit : 'Đơn vị năm'
    },

    EventList : {
        List   : 'Danh sách',
        Start  : 'Bắt đầu',
        Finish : 'Kết thúc'
    },

    DayView : {
        Day               : 'Ngày',
        dayUnit           : 'Đơn vị ngày',
        daysUnit          : 'Đơn vị ngày',
        expandAllDayRow   : 'Mở rộng phần cả ngày',
        collapseAllDayRow : 'Thu gọn phần cả ngày',
        timeFormat        : 'LST'
    },

    Sidebar : {
        'Filter events' : 'Lọc sự kiện'
    },

    WeekExpander : {
        expandTip   : 'Nhấp để mở rộng hàng',
        collapseTip : 'Nhấp để thu gọn hàng'
    }
};

export default LocaleHelper.publishLocale(locale);
